File: 100_SameTree/main.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Iterative_Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {
        if (p == null && q == null) return true;
        if (p == null || q == null) return false;
        Stack<TreeNode> stackP = new Stack<>();
        Stack<TreeNode> stackQ = new Stack<>();
        stackP.push(p);
        stackQ.push(q);
        while (!stackP.empty() && !stackQ.empty()) {
            TreeNode np = stackP.pop();
            TreeNode nq = stackQ.pop();
            if (np.val != nq.val) return false;
            if (np.left != null && nq.left != null) {
                stackP.push(np.left);
                stackQ.push(nq.left);
            } else if (np.left != null || nq.left != null) {
                return false;
            }
            if (np.right != null && nq.right != null) {
                stackP.push(np.right);
                stackQ.push(nq.right);
            } else if (np.right != null || nq.right != null) {
                return false;
            }
        }
        return true;
    }
}

class Recursive_Solution{
    public boolean isSameTree(TreeNode p, TreeNode q) {
        if (p == null && q == null) return true;
        if (q == null || p == null) return false;
        if (p.val != q.val) return false;
        return isSameTree(p.right, q.right) && isSameTree(p.left, q.left);
    }
}

File: 100_SameTree/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 
class Recursive_Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if(p == NULL && q == NULL) return true;
        if(p == NULL || q == NULL) return false;
        if(p->val != q->val) return false;
        return isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
    }
};

// BFS + queue
class Solution2{
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        queue<pair<TreeNode*, TreeNode*>> myQueue;
        myQueue.push(pair<TreeNode*, TreeNode*>(p, q));
        while (!myQueue.empty()) {
            p = myQueue.front().first;
            q = myQueue.front().second;
            if(!p ^ !q || (p && q && p->val != q->val)) break;
            myQueue.pop();
            if(p && q) {
                myQueue.push(pair<TreeNode*, TreeNode*>(p->left, q->left));
                myQueue.push(pair<TreeNode*, TreeNode*>(p->right, q->right));
            }
        }
        return myQueue.empty();
    }
}

